cranelift-codegen/src/isa/x86/abi.rs

```diff
@@ -544,8 +600,20 @@ fn fastcall_prologue_epilogue(func: &mut ir::Function, isa: &dyn TargetIsa) -> C
     func.signature.params.push(fp_arg);
     func.signature.returns.push(fp_arg);
 
-    for csr in csrs.iter(GPR) {
-        let csr_arg = ir::AbiParam::special_reg(reg_type, ir::ArgumentPurpose::CalleeSaved, csr);
+    for gp_csr in csrs.iter(GPR) {
+        let csr_arg = ir::AbiParam::special_reg(reg_type, ir::ArgumentPurpose::CalleeSaved, gp_csr);
+        func.signature.params.push(csr_arg);
+        func.signature.returns.push(csr_arg);
+    }
+
+    for fp_csr in csrs.iter(FPR) {
+        // The calling convention described in
+        // https://docs.microsoft.com/en-us/cpp/build/x64-calling-convention only specifically
+        // discusses XMM6-XMM15, which are 128-bit registers. It may be sufficient to preserve only
+        // the low 128 bits here, but we may find that in practice we should preserve all of
```

Comment:
You only need to preserve the lower 128 bits.
The upper 128 (or 384, whatever) are volatile.
The rationale is:
 - You cannot make registers nonvolatile after the fact.
 - The ABI cannot be revised in a compatible way, adding nonvolatile registers.
 - There are no unwind codes to describe how to restore anything not in the v1 ABI.

You can save more if you want, but it buys little/nothing.

Except, you know, you can have private calling conventions among known callers/callees.

Callers can assume more volatiles (ymm, zmm) are preserved, and can allow for nonvolatiles to be trashed (that the caller preserves).

Calling conventions are largely?entirely about separate compilation.

Reply:
Speaking of unwind, can we update `unwind.rs` to emit the unwind codes for these CSRs as part of this PR?  The original implementation made some assumptions as to what instructions were present in the prologue.

Reply:
I've update Windows fastcall unwind generation to [emit XmmSave128](https://github.com/bytecodealliance/cranelift/pull/1378/files#diff-9fd98f9cfa3e5b06cfaaf0209f9546afR185-R208) and XmmSave128Far, though I'm suspect of the offset.

From reading https://docs.microsoft.com/en-us/cpp/build/exception-handling-x64?view=vs-2019#unwind-operation-code it looks like the offset for SaveXmm counts a number of 16-byte stack slots from the base of the call frame, so storing to `[rbp - 0x10]` would have an offset of 1, `[rbp - 0x20]` would be 2. In contrast, the callee-save code added here should be `[rsp - csr_stack_slot_offset + FPR_offset]` where `FPR_offset` is 0x10, 0x20, ... as appropriate. The offsets recorded are `FPR_offset`, and my suspicion is that they would be interpreted as offsets from the other side of callee-save space by Windows.

This is well beyond my experience with Windows to actually verify. If this looks right to you, I'd be happy to hear it - threading the size of `csr_stack_slot` through to `UnwindInfo` seems kind of annoying. I think it would best be a field on `Function`?

Reply:
I also updated the comment to be more strongly worded on preserving the low 128 bits, thanks for the feedback, @jaykrell.

Reply:
To better understand, just look at C++ examples.
Here is a start. Change it up with alloca to get a frame pointer.
But ouch, keep in mind, compiler does some odd things that seem to violate the spec.
Like getting the rsp/rbp before adjustment, and recording offsets from after.
That does occur below unfortunately -- the use of rax.
That is an optimization and you don't have to do it. Please don't, really.
It is just trying to keep the offsets in the instructions to fit in 8 bits instead of 32 I think, to shrink instruction size, and looks unnecessary in this case, and wastes an instruction.

```
C:\s>type 1.c

__declspec(dllexport)
float (*f1)(float, float);

__declspec(dllexport)
float f2(float a, float b)
{
 float c = a + b;
 float d = a - b;
 float e = a * b;
 float f = a / b;

 return f1(a, b) + f1(c, d) + f1(c, d) - f1(e, f) + f1(e + f, e - f);
}

C:\s>cl /O2 /LD /Zi 1.c /link /noentry /incremental:no
Microsoft (R) C/C++ Optimizing Compiler Version 19.21.27702.2 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

1.c
Microsoft (R) Incremental Linker Version 14.21.27702.2
Copyright (C) Microsoft Corporation.  All rights reserved.

/out:1.dll
/dll
/implib:1.lib
/debug
/noentry
/incremental:no
1.obj
   Creating library 1.lib and object 1.exp

C:\s>link /dump /unwindinfo /disasm 1.dll
Microsoft (R) COFF/PE Dumper Version 14.21.27702.2
Copyright (C) Microsoft Corporation.  All rights reserved.


Dump of file 1.dll

File Type: DLL

f2:
; This instruction is ok but imho an unnecessary and possibly failed optimization.

  0000000180001000: 48 8B C4           mov         rax,rsp

  0000000180001003: 48 81 EC 98 00 00  sub         rsp,98h
                    00
; for example, this could/should be encoded as mov [rsp-80]

  000000018000100A: 0F 29 70 E8        movaps      xmmword ptr [rax-18h],xmm6

; for example, this could/should be encoded as mov [rsp-70]

  000000018000100E: 0F 29 78 D8        movaps      xmmword ptr [rax-28h],xmm7
  0000000180001012: 0F 28 F9           movaps      xmm7,xmm1
  0000000180001015: 44 0F 29 40 C8     movaps      xmmword ptr [rax-38h],xmm8
  000000018000101A: 44 0F 28 C0        movaps      xmm8,xmm0
  000000018000101E: 44 0F 29 48 B8     movaps      xmmword ptr [rax-48h],xmm9
  0000000180001023: 44 0F 28 C8        movaps      xmm9,xmm0
  0000000180001027: 44 0F 29 50 A8     movaps      xmmword ptr [rax-58h],xmm10
  000000018000102C: F3 44 0F 5C CF     subss       xmm9,xmm7
  0000000180001031: 44 0F 29 58 98     movaps      xmmword ptr [rax-68h],xmm11
  0000000180001036: 44 0F 28 D0        movaps      xmm10,xmm0
  000000018000103A: 44 0F 29 60 88     movaps      xmmword ptr [rax-78h],xmm12
  000000018000103F: F3 44 0F 58 D7     addss       xmm10,xmm7
  0000000180001044: 44 0F 28 E0        movaps      xmm12,xmm0
  0000000180001048: 44 0F 28 D8        movaps      xmm11,xmm0
  000000018000104C: F3 44 0F 59 E7     mulss       xmm12,xmm7
  0000000180001051: 41 0F 28 C9        movaps      xmm1,xmm9
  0000000180001055: 41 0F 28 C2        movaps      xmm0,xmm10
  0000000180001059: F3 44 0F 5E DF     divss       xmm11,xmm7
  000000018000105E: FF 15 A4 1F 00 00  call        qword ptr [f1]
  0000000180001064: 0F 28 F0           movaps      xmm6,xmm0
  0000000180001067: 0F 28 CF           movaps      xmm1,xmm7
  000000018000106A: 41 0F 28 C0        movaps      xmm0,xmm8
  000000018000106E: FF 15 94 1F 00 00  call        qword ptr [f1]
  0000000180001074: F3 0F 58 F0        addss       xmm6,xmm0
  0000000180001078: 41 0F 28 C9        movaps      xmm1,xmm9
  000000018000107C: 41 0F 28 C2        movaps      xmm0,xmm10
  0000000180001080: FF 15 82 1F 00 00  call        qword ptr [f1]
  0000000180001086: 0F 28 F8           movaps      xmm7,xmm0
  0000000180001089: 41 0F 28 CB        movaps      xmm1,xmm11
  000000018000108D: 41 0F 28 C4        movaps      xmm0,xmm12
  0000000180001091: F3 0F 58 FE        addss       xmm7,xmm6
  0000000180001095: FF 15 6D 1F 00 00  call        qword ptr [f1]
  000000018000109B: 41 0F 28 CC        movaps      xmm1,xmm12
  000000018000109F: F3 0F 5C F8        subss       xmm7,xmm0
  00000001800010A3: F3 41 0F 5C CB     subss       xmm1,xmm11
  00000001800010A8: F3 45 0F 58 DC     addss       xmm11,xmm12
  00000001800010AD: 41 0F 28 C3        movaps      xmm0,xmm11
  00000001800010B1: FF 15 51 1F 00 00  call        qword ptr [f1]

 ; blech, see below 
  00000001800010B7: 4C 8D 9C 24 98 00  lea         r11,[rsp+98h]
                    00 00
  00000001800010BF: 41 0F 28 73 E8     movaps      xmm6,xmmword ptr [r11-18h]
  00000001800010C4: F3 0F 58 C7        addss       xmm0,xmm7
  00000001800010C8: 0F 28 7C 24 70     movaps      xmm7,xmmword ptr [rsp+70h]
  00000001800010CD: 45 0F 28 43 C8     movaps      xmm8,xmmword ptr [r11-38h]
  00000001800010D2: 45 0F 28 4B B8     movaps      xmm9,xmmword ptr [r11-48h]
  00000001800010D7: 45 0F 28 53 A8     movaps      xmm10,xmmword ptr [r11-58h]
  00000001800010DC: 45 0F 28 5B 98     movaps      xmm11,xmmword ptr [r11-68h]
  00000001800010E1: 45 0F 28 63 88     movaps      xmm12,xmmword ptr [r11-78h]

; This violates the letter of the spec, alas.
; It is supposed to be add rsp,n, or lea rsp, n[frame]
  00000001800010E6: 49 8B E3           mov         rsp,r11
  00000001800010E9: C3                 ret

Function Table (1)

           Begin    End      Info      Function Name

  00000000 00001000 000010EA 0000211C  f2
    Unwind version: 1
    Unwind flags: None
    Size of prologue: 0x3F
    Count of codes: 16
    Unwind codes:
      3F: SAVE_XMM128, register=xmm12 offset=0x20
      36: SAVE_XMM128, register=xmm11 offset=0x30
      2C: SAVE_XMM128, register=xmm10 offset=0x40
      23: SAVE_XMM128, register=xmm9 offset=0x50
      1A: SAVE_XMM128, register=xmm8 offset=0x60
      12: SAVE_XMM128, register=xmm7 offset=0x70
      0E: SAVE_XMM128, register=xmm6 offset=0x80
      0A: ALLOC_LARGE, size=0x98

```

So, overall, mixed. The C++ examples help to understand, but they also confuse, because the C++ compiler violates the spec in ways that end up meaning the same thing at runtime, if you understand closely how unwinding works.
I would encourage you to follow the spec instead, though I suppose there might be measurable perf benefits otherwise.

You can do the link /dump /unwindinfo on any binary -- no need for symbols.

Reply:
The NT unwinder is also open sourced, here:

https://github.com/dotnet/runtime/blob/master/src/coreclr/src/unwinder/amd64/unwinder_amd64.cpp#L895

at least an old maybe subsetted/altered form, but probably useful.

Reply:
The offsets in the generated unwind information in the test do seem odd, especially compared to the above output form VC++, but I think that's a result of adjusting the stack pointer for the frame *after* we're writing in the SIMD CSRs.  In the test output, an offset of 0, 0x10, and 0x20 make sense given that the SP will be adjusted up during unwinding restores the XMM registers.

Perhaps we should adjust the SP first?

Reply:
If the generated unwind info looks okay, are you good with this change and revisiting in case it turns out not to be? I don't have anything resembling a useful Windows development environment, so I'm not even entirely sure what to look for being wrong in case the unwind information is wrong - I imagine "only" an incorrect value for a callee-save register in some parent frame? Given that overall this PR fixes an ABI error that results in dangerously wrong code, I _would_ like to get this figured out :)

Reply:
I'm good with the change and open to revisiting it in the future.  I think the order of operations in our prologues is just a little "strange" compared to VC++'s prologues, but I don't necessarily see that as a problem with respect to unwind.

Reply:
The VIsual C++ code above is weird.
I do not recommend emulating it.
It is surely correct, and it might be optimized, but do not emulate it.
Visual C++ definitely does some strange things as optimizations.
In particular, it stores to the stack before adjusting rsp, and then has to misstate frame pointer offsets. "Normal" code only pushes prior to storing to the stack, I believe.
Maybe try looking at unoptimized output instead.
Or link /dump /disasm /unwindinfo ntoskrnl.exe to find simple stuff written in assembly.
Or read and follow the spec -- Visual C++ kinda skirts the spec.